### ISSUE TITLE
Flow inputs annotated as container types crash with Pydantic V2

### ISSUE THREAD

**Issue by psktam** [PrefectHQ/prefect#11149]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

If I define a flow that has a typed container as an input, prefect will crash when it tries to deploy that flow using the `flow.serve` function. Downgrading pydantic to a version below major version 2 fixes the issue, but our project requires pydantic == 2.2.1

It looks like this issue might be related to this issue? https://github.com/PrefectHQ/prefect/issues/10506 but they're running a lower pydantic version, which should have worked here.

### Reproduction

```python3
from prefect import flow


@flow
def a_flow(arg1: list[str]) -> str:
    print(arg1)
    return ",".join(arg1)


if __name__ == "__main__":
    a_flow.serve(name="test-flow")
```

### Error

```python3
Traceback (most recent call last):
  File "/Users/big-n00b/local-repo/run_flow.py", line 5, in <module>
    def mini_flow(arg1: list[str]) -> str:
  File "/Users/big-n00b/Library/Caches/pypoetry/virtualenvs/my-project-RcQVLUoH-py3.10/lib/python3.10/site-packages/prefect/flows.py", line 1339, in flow
    Flow(
  File "/Users/big-n00b/Library/Caches/pypoetry/virtualenvs/my-project-RcQVLUoH-py3.10/lib/python3.10/site-packages/prefect/context.py", line 185, in __register_init__
    __init__(__self__, *args, **kwargs)
  File "/Users/big-n00b/Library/Caches/pypoetry/virtualenvs/my-project-RcQVLUoH-py3.10/lib/python3.10/site-packages/prefect/flows.py", line 299, in __init__
    self.parameters = parameter_schema(self.fn)
  File "/Users/big-n00b/Library/Caches/pypoetry/virtualenvs/my-project-RcQVLUoH-py3.10/lib/python3.10/site-packages/prefect/utilities/callables.py", line 322, in parameter_schema
    if HAS_PYDANTIC_V2 and has_v2_model_as_param(signature):
  File "/Users/big-n00b/Library/Caches/pypoetry/virtualenvs/my-project-RcQVLUoH-py3.10/lib/python3.10/site-packages/prefect/_internal/pydantic/v2_schema.py", line 22, in has_v2_model_as_param
    return any(
  File "/Users/big-n00b/Library/Caches/pypoetry/virtualenvs/my-project-RcQVLUoH-py3.10/lib/python3.10/site-packages/prefect/_internal/pydantic/v2_schema.py", line 24, in <genexpr>
    or (inspect.isclass(p.annotation) and issubclass(p.annotation, V2BaseModel))
  File "/Users/big-n00b/.pyenv/versions/3.10.12/lib/python3.10/abc.py", line 123, in __subclasscheck__
    return _abc_subclasscheck(cls, subclass)
TypeError: issubclass() arg 1 must be a class
```

### Versions

```Text
Version:       2.14.3
API version:     0.8.4
Python version:   3.10.12
Git commit:     f1ff9257
Built:        Thu, Nov 2, 2023 4:12 PM
OS/Arch:       darwin/arm64
Profile:       default
```

**Comment by serinamarie** [comment]
Hi, @psktam, thanks for creating your first issue with Prefect! I am able to replicate your issue, as we only test Pydantic V2 compatibility with our highest/lowest Python versions (3.8 and 3.11). Are you able to upgrade to 3.11 by chance?

**Comment by serinamarie** [comment]
I've also opened a [draft PR](https://github.com/PrefectHQ/prefect/pull/11150/files) to update our workflow so that we do test again 3.10 and 3.9 in future.

**Comment by urimandujano** [comment]
@psktam, thanks for bringing this to our attention. I believe this is due to how python handled types in versions <3.11 and reflected a gap in our tests.[ We've added some code to handle errors a more gracefully and added tests verifying this fix](https://github.com/PrefectHQ/prefect/pull/11152). 

Look out for our next release which will ship this fix. In the meantime, I was able to work around this issue by adding this to the top of my flow test file:

```python
from __future__ import annotations
```
